dcpy/metadata/validate.py

```diff
@@ -0,0 +1,158 @@
+import pandas as pd
+from pathlib import Path
+import shapely
+import yaml
+
+
+def validate_csv(csv_path: Path, metadata_path: Path):
+    """Validate a csv against a metadata file."""
+    csv_df = pd.read_csv(csv_path, dtype=str)
+    csv_df_non_null = csv_df.fillna("")
+
+    metadata = yaml.safe_load(open(metadata_path, "r"))
+
+    errors = []
+
+    # Find mismatched columns
+    # Assumes that if `appears_in` isn't set, every filetype will contain the column
+    metadata_column_names = {
+        x["name"]
+        for x in metadata["columns"]
+        if not x.get("appears_in") or "csv" in x.get("appears_in")
+    }
+    csv_headers = set(csv_df.columns)
+
+    extras_in_csv = csv_headers.difference(metadata_column_names)
+    if extras_in_csv:
+        errors.append(
+            [
+                "COLUMN_MISMATCH",
+                f"Invalid column(s) found in source data: {extras_in_csv}.",
+            ]
+        )
+
+    not_found_in_csv = metadata_column_names.difference(csv_headers)
+    if not_found_in_csv:
+        errors.append(
+            [
+                "COLUMN_MISMATCH",
+                f"Column(s) missing from source data: {not_found_in_csv}.",
+            ]
+        )
+
+    # Validate Data in Columns
+    for col in metadata["columns"]:
+        col_name = col["name"]
+        if col_name in not_found_in_csv:
+            continue
+
+        match col["data_type"]:
+            case "bbl":
+                with_invalid_bbl = csv_df[~csv_df["BBL"].str.match("^\d{10}$")]
+                if not with_invalid_bbl.empty:
```

Comment:
could `with_invalid_bbl` be `invalids` like the other data types?

Reply:
Sure could. Will change in followup